sessionctx/variable/sysvar.go

```diff
@@ -1628,6 +1628,77 @@ var defaultSysVars = []*SysVar{
 	{Scope: ScopeGlobal, Name: TiDBGCLifetime, Value: "10m0s", Type: TypeDuration, MinValue: int64(time.Minute * 10), MaxValue: math.MaxInt64},
 	{Scope: ScopeGlobal, Name: TiDBGCConcurrency, Value: "-1", Type: TypeInt, MinValue: 1, MaxValue: 128, AllowAutoValue: true},
 	{Scope: ScopeGlobal, Name: TiDBGCScanLockMode, Value: "PHYSICAL", Type: TypeEnum, PossibleValues: []string{"PHYSICAL", "LEGACY"}},
+
+	// variable for top SQL feature.
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBEnableTopSQL, Value: BoolToOnOff(config.DefTopSQLEnable), Type: TypeBool, AllowEmpty: true, GetSession: func(s *SessionVars) (string, error) {
+		return BoolToOnOff(config.GetGlobalConfig().TopSQL.Enable), nil
+	}, SetSession: func(vars *SessionVars, s string) error {
+		config.UpdateGlobal(func(conf *config.Config) {
+			conf.TopSQL.Enable = TiDBOptOn(s)
+		})
+		return nil
+	}, SetGlobal: func(vars *SessionVars, s string) error {
+		config.UpdateGlobal(func(conf *config.Config) {
```

Comment:
> GetGlobalConfig and StoreGlobalConfig is atomic, but the returned pointer of config is not guarded by the atomic. newConf := *g is not race free.

As mentioned in https://github.com/pingcap/tidb/issues/23748, `UpdateGlobal` is not thread-safe and should only be used in serial test.

Reply:
Addressed.